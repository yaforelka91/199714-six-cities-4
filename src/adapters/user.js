const adaptUser = (user) => {
  return {
    id: user.id,
    email: user.email,
    name: user.name,
    picture: user.avatar_url,
    isSuper: user.is_pro,
  };
};

export default adaptUser;
