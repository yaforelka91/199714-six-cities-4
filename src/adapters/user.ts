import User from '../interfaces/user';

const adaptUser = (user): User => ({
  id: user.id,
  email: user.email,
  name: user.name,
  picture: user.avatar_url,
  isSuper: user.is_pro,
});

export default adaptUser;
