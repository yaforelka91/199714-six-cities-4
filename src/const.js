export const OfferType = {
  APARTMENT: `Apartment`,
  ROOM: `Private room`,
  HOUSE: `House`,
  HOTEL: `Hotel`,
};
