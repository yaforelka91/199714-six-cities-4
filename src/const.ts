export const MONTHS_LIST = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export enum SortingTypes {
  POPULAR = `popular`,
  TO_HIGH = `to-high`,
  TO_LOW = `to-low`,
  TOP_RATED = `top-rated`,
};

export enum IconSizes {
  WIDTH = 27,
  HEIGHT = 39
};

export enum AppRoute {
  FAVORITES = `/favorites`,
  LOGIN = `/login`,
  OFFER = `/offer`,
  ROOT = `/`
};

export enum CardView {
  CITIES = `cities`,
  NEAR = `near`,
  FAVORITES = `favorites`
};
