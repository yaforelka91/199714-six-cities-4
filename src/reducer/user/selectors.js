import NameSpace from '../name-space.js';

const NAME_SPACE = NameSpace.USER;

export const getAuthorizationStatus = (state) => {
  return state[NAME_SPACE].authorizationStatus;
};

export const getAuthorizationProgress = (state) => {
  return state[NAME_SPACE].isAuthorizationInProgress;
};

export const getUserData = (state) => {
  return state[NAME_SPACE].userData;
};

export const getError = (state) => {
  return state[NAME_SPACE].errorType;
};
