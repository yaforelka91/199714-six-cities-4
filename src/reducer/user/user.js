import {extend} from '../../utils.js';
import adaptUser from '../../adapters/user.js';
import adaptError from '../../adapters/error.js';
import {Operation as DataOperation} from '../data/data.js';
import {Error} from '../../api.js';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  userData: {},
  isAuthorizationInProgress: true,
  errorType: ``,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_USER_DATA: `SET_USER_DATA`,
  CHANGE_PROGRESS_STATUS: `CHANGE_PROGRESS_STATUS`,
  CATCH_ERROR: `CATCH_ERROR`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  setUserData: (authData) => {
    return {
      type: ActionType.SET_USER_DATA,
      payload: authData,
    };
  },
  changeProgressStatus: (isLoaded) => {
    return {
      type: ActionType.CHANGE_PROGRESS_STATUS,
      payload: isLoaded,
    };
  },
  catchError: (errorMessage) => {
    return {
      type: ActionType.CATCH_ERROR,
      payload: errorMessage,
    };
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_USER_DATA:
      return extend(state, {
        userData: action.payload,
      });

    case ActionType.CHANGE_PROGRESS_STATUS:
      return extend(state, {
        isAuthorizationInProgress: action.payload,
      });

    case ActionType.CATCH_ERROR:
      return extend(state, {
        errorType: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setUserData(adaptUser(response.data)));
      })
      .catch((err) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        throw err;
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setUserData(adaptUser(response.data)));
        dispatch(ActionCreator.catchError(``));
        dispatch(DataOperation.loadOffers());
      })
      .catch((err) => {
        dispatch(ActionCreator.changeProgressStatus(false));

        if (err.response.status === Error.BAD_REQUEST) {
          dispatch(ActionCreator.catchError(adaptError(err.response.data.error)));
          return;
        }

        throw err;
      });
  },
};

export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
