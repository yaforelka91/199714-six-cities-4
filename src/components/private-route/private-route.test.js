import React from 'react';
import renderer from 'react-test-renderer';
import PrivateRoute from './private-route.jsx';
import {AuthorizationStatus} from '../../reducer/user/user.js';
import {MemoryRouter} from 'react-router';
import {AppRoute} from '../../const.js';

const MockComponent = () => <div />;

describe(`PrivateRouteSnapshot`, () => {
  it(`should render Loader`, () => {
    const tree = renderer.create(
        <MemoryRouter initialEntries={[AppRoute.FAVORITES]}>
          <PrivateRoute
            exact
            path={AppRoute.FAVORITES}
            authorizationStatus={AuthorizationStatus.AUTH}
            isAuthorizationInProgress={true}
            render={() => {
              return <MockComponent />;
            }}
          />
        </MemoryRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render PivateRoute with mock component`, () => {
    const tree = renderer.create(
        <MemoryRouter initialEntries={[AppRoute.FAVORITES]}>
          <PrivateRoute
            exact
            path={AppRoute.FAVORITES}
            authorizationStatus={AuthorizationStatus.AUTH}
            render={() => {
              return <MockComponent />;
            }}
          />
        </MemoryRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
