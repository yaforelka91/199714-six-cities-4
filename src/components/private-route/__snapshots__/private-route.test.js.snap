// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRouteSnapshot should render Loader 1`] = `
<p>
  Check autorization...
</p>
`;

exports[`PrivateRouteSnapshot should render PivateRoute with mock component 1`] = `<div />`;
